### PR TITLE
Add general-purpose serving infrastructure

### PR DESCRIPTION
**Pull request by sethvargo**
This standardizes our http/grcp server creation and lifecycle handling.

### CODE REVIEW

File: serving/serving.go [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+		if err := srv.Serve(s.listener); err != nil && !errors.Is(err, http.ErrServerClosed) {
+			select {
+			case errCh <- err:
+			default:
+			}
+		}
+	}()
+
+	// Wait for the provided context to finish.
+	<-ctx.Done()
```
Comment (yolocs) [inline serving/serving.go:123]: If the context is not done but `Serve` func returns an error, would this unintentionally block the code? Do we even want to block if there is an error?

File: serving/serving.go [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+		if err := srv.Serve(s.listener); err != nil && !errors.Is(err, grpc.ErrServerStopped) {
+			select {
+			case errCh <- err:
+			default:
+			}
+		}
+	}()
+
+	// Wait for the provided context to finish.
+	<-ctx.Done()
```
Comment (yolocs) [inline serving/serving.go:190]: Same question

File: serving/serving.go (hunk 1 above)
Comment (sethvargo) [reply serving/serving.go]: Oh yea, that would be a bad bug 😄. Fixed - thanks for catching.

File: serving/serving.go [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+	logger := logging.FromContext(ctx)
+
+	// Start the server in a background goroutine so we can listen for cancellation
+	// in the main process.
+	errCh := make(chan error, 1)
+	doneCh := make(chan struct{})
+	go func() {
+		defer close(doneCh)
+
+		logger.Infow("server is starting", "ip", s.ip, "port", s.port)
```
Comment (pdewilde) [inline serving/serving.go:111]: I'm curious, why use the non-sentence convention for log messages? Wouldn't it make sense for these?

**Review by yolocs — APPROVED** [review]

**Comment by pdewilde** [comment]
Another question: would it make sense to push this out to sandwich?

File: serving/serving.go [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+// Note that the incoming [net/http.Server]'s address is ignored over the
+// listener's configuration.
+func (s *Server) StartHTTP(ctx context.Context, srv *http.Server) error {
+	logger := logging.FromContext(ctx)
+
+	// Start the server in a background goroutine so we can listen for cancellation
+	// in the main process.
+	errCh := make(chan error, 1)
+	doneCh := make(chan struct{})
+	go func() {
```
[the remark this answers is not in the document]
Comment (sethvargo) [reply serving/serving.go:108]: Yea - I had that thought and went down that path. It ends up adding a layer of indirection that felt more harmful than helpful.

**Comment by sethvargo** [comment]
> Another question: would it make sense to push this out to sandwich?
> 
> Or is there value there in having less indirection for someone new to the codebase?

It depends on how "production ready" you want the examples in sandwhich to be.

**Comment by pjh** [comment]
> It depends on how "production ready" you want the examples in sandwhich to be.

Can you elaborate? We intend to write production-ready code for Sandwich, though we're getting there iteratively.

It's not clear to me if the interface in this PR is something we'll want to adopt or even be able to adopt in Sandwich (given our use of Chi there), but from the PR description it sounds like we probably should.

**Comment by sethvargo** [comment]
> It's not clear to me if the interface in this PR is something we'll want to adopt or even be able to adopt in Sandwich (given our use of Chi there), but from the PR description it sounds like we probably should.

This is for _serving_, Chi is for _routing_. Chi implements `http.Handler`, so a Chi instance could be passed to `ServeHTTPHandler` or `ServeHTTP` with an `http.Server#Handler`.

**Pull request merged**